src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs

```diff
@@ -196,6 +194,37 @@ BoundDecisionDagNode makeReplacement(BoundDecisionDagNode dag, Func<BoundDecisio
             }
         }
 
+        /// <summary>
+        /// To facilitate subsumption-checking for list-patterns, some nodes may be inserted into the dag during construction.
+        /// We will remove such nodes immediately after as these are not important in lowering or any other analysis.
+        /// </summary>
+        public BoundDecisionDag RemoveCompilerGeneratedNodes()
+        {
+            return Rewrite(static (node, replacement) =>
+            {
+                if (node.WasCompilerGenerated)
```

Comment:
>node.WasCompilerGenerated [](http://example.com/codeflow?start=20&length=25)

Relying on WasCompilerGenerated feels fragile in the long term.

Reply:
I'm not sure if I understand the concern. I imagine WasCompilerGenerated is used for all kinds of reasons, and in the dag builder, it's used for this particular purpose only. I don't see how a separate flag helps while WasCompilerGenerated is just there already (and given the name, it's not "incorrect" to use it)

Note that we also propagated the flag to the containing dag node for a simpler check in this method which is called for every decision dag so I'd rather we keep it as simple as possible.

Reply:
> I'm not sure if I understand the concern. I imagine WasCompilerGenerated is used for all kinds of reasons, and in the dag builder, it's used for this particular purpose only. I don't see how a separate flag helps while WasCompilerGenerated is just there already (and given the name, it's not "incorrect" to use it)

This flag has very specific meaning, the node represents an operation that was not explicitly "mentioned" in the original syntax. Its primary purpose is to inform SemanticModel. It was never meant to represent the fact that the node can be ignored/eliminated. I cannot remember any example where an addition of the fag changes the executable code that we emit. In that sense, it is always safe to add a flag like that. I guess it is fine to set the flag for the nodes, if you'd like. However, "overriding" the meaning of the flag in one component feels fragile and error prone.

Reply:
Thanks for the explanation. I think dag nodes aren't exposed through semantic model, so I'd guess it's not harmful to make use of this flag. Just to demonstrate, here's how I would do it without using WasCompilerGenerated. 
```cs
switch (node)
{
    case BoundEvaluationDecisionDagNode { Evaluation: BoundDagAssignmentEvaluation } evalNode:
        return replacement[evalNode.Next];
    case BoundTestDecisionDagNode { Test: BoundDagValueTest { IsSubsumptionOnly: true } } testNode:
        return replacement[testNode.WhenFalse];
    default:
        return TrivialReplacement(node, replacement);
}
```
That seems to be too much unnecessary work. But again, if not a concern, I'll do that instead.

Reply:
> That seems to be too much unnecessary work.

I don't see much difference by comparison to the code in commit 6 in terms of what the code is doing. However, the intent is much clear, which is a good thing.